src/core/drive/morph_renderer.js

```diff
@@ -0,0 +1,92 @@
+import Idiomorph from "idiomorph"
+import { dispatch, nextAnimationFrame } from "../../util"
+import { Renderer } from "../renderer"
+
+export class MorphRenderer extends Renderer {
+  async render() {
+    if (this.willRender) await this.#morphBody()
+  }
+
+  get renderMethod() {
+    return "morph"
+  }
+
+  // Private
+
+  async #morphBody() {
+    this.#morphElements(this.currentElement, this.newElement)
+    this.#reloadRemoteFrames()
+
+    dispatch("turbo:morph", {
+      detail: {
+        currentElement: this.currentElement,
+        newElement: this.newElement
+      }
+    })
+  }
+
+  #morphElements(currentElement, newElement, morphStyle = "outerHTML") {
+    this.isMorphingTurboFrame = this.#isFrameReloadedWithMorph(currentElement)
+
+    Idiomorph.morph(currentElement, newElement, {
+      morphStyle: morphStyle,
+      callbacks: {
+        beforeNodeMorphed: this.#shouldMorphElement,
+        beforeNodeRemoved: this.#shouldRemoveElement,
+        afterNodeMorphed: this.#reloadStimulusControllers
+      }
+    })
+  }
+
+  #reloadRemoteFrames() {
+    this.#remoteFrames().forEach((frame) => {
+      if (this.#isFrameReloadedWithMorph(frame)) {
+        this.#renderFrameWithMorph(frame)
+        frame.reload()
+      }
+    })
+  }
+
+  #renderFrameWithMorph(frame) {
+    frame.addEventListener("turbo:before-frame-render", (event) => {
+      event.detail.render = this.#morphFrameUpdate
+    }, { once: true })
+  }
+
+  #morphFrameUpdate = (currentElement, newElement) => {
+    dispatch("turbo:before-frame-morph", {
+      target: currentElement,
+      detail: { currentElement, newElement }
+    })
+    this.#morphElements(currentElement, newElement)
+  }
```

Comment:
Would it be possible to achieve this branch's goals without introducing new types of events?

Could a morphing-supported `<turbo-frame>` dispatch `turbo:before-frame-render` and `turbo:frame-render` events with `renderMethod: "morph"` in the same way we're dispatching `turbo:render` events?

When handling `turbo:before-frame-render` events with `renderMethod: "morph"`, listeners can access the `currentElement` analog with `event.target` and the `newElement` analog with `event.detail.newFrame`.

I also wonder if would-be `turbo:morph` listeners could instead listen for `turbo:render` events with `renderMethod: "morph"`, then access the `currentElement` from `event.target` and the `newElement` from `event.detail.newBody`.